File: WhirlyGlobeSrc/WhirlyGlobeComponentTester/WhirlyGlobeComponentTester/AnimationTest.m
//
//  AnimationTest.m
//  WhirlyGlobeComponentTester
//
//  Created by Steve Gifford on 7/31/13.
//  Copyright (c) 2013 mousebird consulting. All rights reserved.
//

#import "AnimationTest.h"
//#import "WhirlyGlobe.h"

// Note: Should import WhirlyGlobe.h here instead of doing this
//       But we can hack it for the moment

@implementation AnimatedSphere
{
    NSTimeInterval start;
    float period;
    float radius;
    UIColor *color;
    MaplyCoordinate startPt;
    MaplyComponentObject *sphereObj;
}

- (id)initWithPeriod:(float)inPeriod radius:(float)inRadius color:(UIColor *)inColor viewC:(MaplyBaseViewController *)viewC
{
    self = [super initWithViewController:viewC];
    if (!self)
        return nil;
    
    period = inPeriod;
    radius = inRadius;
    color = inColor;
    
    start = CFAbsoluteTimeGetCurrent();
    return self;
}

- (bool)hasUpdate
{
    return true;
}

//- (void)updateForFrame:(WhirlyKit::RendererFrameInfo *)frameInfo
- (void)updateForFrame:(id)frameInfo
{
    if (sphereObj)
    {
        [super.viewC removeObjects:@[sphereObj] mode:MaplyThreadCurrent];
        sphereObj = nil;
    }

    float t = (CFAbsoluteTimeGetCurrent()-start)/period;
    t -= (int)t;
    
    MaplyCoordinate center = MaplyCoordinateMakeWithDegrees(-180+t*360.0, 0.0);
    
    MaplyShapeSphere *sphere = [[MaplyShapeSphere alloc] init];
    sphere.radius = radius;
    sphere.center = center;
    
    // Here's the trick, we must use MaplyThreadCurrent to make this happen right now
    sphereObj = [super.viewC addShapes:@[sphere] desc:@{kMaplyColor: color} mode:MaplyThreadCurrent];
}

- (void)shutdown
{
    if (sphereObj)
    {
        [super.viewC removeObjects:@[sphereObj] mode:MaplyThreadCurrent];
        sphereObj = nil;
    }
}


@end

File: WhirlyGlobeSrc/WhirlyGlobeSwiftTester/WhirlyGlobeSwiftTester/AnimationTest.swift
//
//  AnimationTest.swift
//  WhirlyGlobeSwiftTester
//
//  Created by jmnavarro on 29/09/15.
//  Copyright (c) 2015 Mousebird. All rights reserved.
//

import Foundation

/** The animation test object runs a sphere around the globe
over a defined time period.
*/
class AnimatedSphere: MaplyActiveObject {

	private let start: NSTimeInterval
	private let period: Float
	private let radius: Float
	private let color: UIColor
	private var sphereObj: MaplyComponentObject?


	/// Initialize with period (amount of time for one orbit), radius and color of the sphere and a starting point
	init(period: Float, radius: Float, color: UIColor, viewC: MaplyBaseViewController) {
		self.period = period
		self.radius = radius
		self.color = color

		start = CFAbsoluteTimeGetCurrent()

		super.init(viewController: viewC)
	}

	var hasUpdate: Bool {
		return true
	}

	func updateForFrame(frameInfo: AnyObject?) {
		if let sphereObj = sphereObj {
			viewC?.removeObjects([sphereObj], mode: .Current)
			self.sphereObj = nil
		}

		let t = (CFAbsoluteTimeGetCurrent()-start)/Double(period)

		let center = MaplyCoordinateMakeWithDegrees(Float(-180+t*360.0), Float(0.0))

		let sphere = MaplyShapeSphere()
		sphere.radius = radius
		sphere.center = center

		// Here's the trick, we must use MaplyThreadCurrent to make this happen right now
		sphereObj = viewC!.addShapes([sphere], desc: [kMaplyColor: color], mode: .Current)
	}

	func shutdown() {
		if let sphereObj = sphereObj {
			viewC?.removeObjects([sphereObj], mode: .Current)
			self.sphereObj = nil
		}
	}

}
